test/e2e/helpers.js

```diff
@@ -147,9 +147,46 @@ async function withFixtures(options, testSuite) {
   }
 }
 
+const getWindowHandles = async (driver) => {
+  await driver.waitUntilXWindowHandles(3);
+  const windowHandles = await driver.getAllWindowHandles();
+
+  const extension = windowHandles[0];
+  const dapp = await driver.switchToWindowWithTitle(
+    'E2E Test Dapp',
+    windowHandles,
+  );
+  const popup = windowHandles.find(
+    (handle) => handle !== extension && handle !== dapp,
+  );
+  return { extension, dapp, popup };
+};
+
+const connectDappWithExtensionPopup = async (driver) => {
+  await driver.openNewPage('http://127.0.0.1:8080/');
```

Comment:
Any way to tie this to starting the dapp server? It seems like you couldn't call this function without first doing that. Maybe this function should at least take the port number for the dapp server in case that ever changes?

Reply:
dapp is initialised [here](https://github.com/MetaMask/metamask-extension/blob/develop/test/e2e/helpers.js#L20). I did the code change to re-use port constant here. Ideally it will be better to combine code to initialise and connect dapp, but that would need a bigger refactor in many test cases.